nhaystack-rt/src/nhaystack/server/tags/BNSiteRefRelation.java

```diff
@@ -56,24 +58,32 @@
      * BControlPoint.
      */
     @Override
-    public Optional<Relation> getRelation(Entity point)
+    public Optional<Relation> getRelation(Entity source)
     {
-        if (!(point instanceof BControlPoint))
+        if (!(source instanceof BControlPoint))
         {
             return Optional.empty();
         }
 
         Id relationId = getRelationId();
-        if (!hasDirectRelation((BComponent) point, relationId))
+        if (hasDirectRelation((BComponent) source, relationId))
         {
-            BComponent equip = getEquip((BControlPoint) point);
-            if (equip != null)
+            return Optional.empty();
+        }
+
+        BComponent equip = getEquip((BControlPoint) source);
+        if (equip == null)
+        {
+            return Optional.empty();
+        }
+
+        Optional<Relation> equipSiteRef = equip.relations().get(relationId, Relations.OUT);

```

Comment:
This was the big performance fix: specifying the relation ID drastically reduces the number of tag rules evaluated to imply the relation.